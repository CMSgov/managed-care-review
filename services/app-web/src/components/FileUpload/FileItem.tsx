import React from 'react'
import { Button } from '@trussworks/react-uswds'
import classnames from 'classnames'
import { SPACER_GIF } from './constants'

import styles from './FileUpload.module.scss'

export type FileStatus =
    | 'DUPLICATE_NAME_ERROR'
    | 'PENDING'
    | 'UPLOAD_COMPLETE'
    | 'UPLOAD_ERROR'
    | 'SAVED_TO_SUBMISSION'

// TODO: May want to use a union data type here to further clarify states
// FileItemUploadComplete | FileItemUploadIncomplete
export type FileItemT = {
    id: string
    name: string
    file?: File // only items that are not uploaded to s3 have this
    url?: string // only items uploaded to s3 have this
    key?: string // only items uploaded to s3 have this
    s3URL?: string // only items uploaded to s3 have this
    status: FileStatus
}

export type FileItemProps = {
    item: FileItemT
    deleteItem: (item: FileItemT) => void
    retryItem: (item: FileItemT) => void
}
export const FileItem = ({
    item,
    deleteItem,
    retryItem,
}: FileItemProps): React.ReactElement => {
    const { name, status } = item
    const hasDuplicateNameError = status === 'DUPLICATE_NAME_ERROR'
    const hasUploadError = status === 'UPLOAD_ERROR'
    const isLoading = status === 'PENDING'
    const isPDF = name.indexOf('.pdf') > 0
    const isWord = name.indexOf('.doc') > 0 || name.indexOf('.pages') > 0
    const isVideo = name.indexOf('.mov') > 0 || name.indexOf('.mp4') > 0
    const isExcel = name.indexOf('.xls') > 0 || name.indexOf('.numbers') > 0
    const isGeneric = !isPDF && !isWord && !isVideo && !isExcel

    const imageClasses = classnames('usa-file-input__preview-image', {
        'is-loading': isLoading,
        'usa-file-input__preview-image--pdf': isPDF,
        'usa-file-input__preview-image--word': isWord,
        'usa-file-input__preview-image--video': isVideo,
        'usa-file-input__preview-image--excel': isExcel,
        'usa-file-input__preview-image--generic': isGeneric,
    })

    const handleDelete = (e: React.MouseEvent) => {
        deleteItem(item)
    }

    const handleRetry = (e: React.MouseEvent) => {
        retryItem(item)
    }

    const Error = (): React.ReactElement | null => {
        if (hasDuplicateNameError)
            return (
                <>
                    <span className={styles.fileItemErrorMessage}>
                        Duplicate file
                    </span>
                    <span className={styles.fileItemErrorMessage}>
                        Please remove
                    </span>
                </>
            )
        else if (hasUploadError)
            return (
                <>
                    <span className={styles.fileItemErrorMessage}>
                        Upload failed
                    </span>
                    <span className={styles.fileItemErrorMessage}>
                        Please remove or retry
                    </span>
                </>
            )
        else {
            return null
        }
    }

    return (
        <>
            <div className={styles.fileItemText}>
                <img
                    id={item.id}
                    data-testid="file-input-preview-image"
                    src={SPACER_GIF}
                    alt=""
                    className={imageClasses}
                />
                <span
                    style={{
                        display: 'flex',
                        flexDirection: 'column',
                        fontSize: 'inherit',
                    }}
                >
                    <Error />
                    <span>{name}</span>
                </span>
            </div>
            <div className={styles.fileItemButtons}>
                <Button
                    type="button"
                    size="small"
                    unstyled
                    onClick={handleDelete}
                >
                    Remove
                </Button>
                {hasUploadError && (
                    <Button
                        type="button"
                        size="small"
                        unstyled
                        onClick={handleRetry}
                    >
                        Retry
                    </Button>
                )}
            </div>
        </>
    )
}
