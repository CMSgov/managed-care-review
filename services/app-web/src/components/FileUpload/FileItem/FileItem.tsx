import React from 'react'
import { Button } from '@trussworks/react-uswds'
import classnames from 'classnames'
import { SPACER_GIF } from '../constants'

import styles from '../FileUpload.module.scss'

export type FileStatus =
    | 'DUPLICATE_NAME_ERROR'
    | 'PENDING'
    | 'SCANNING'
    | 'SCANNING_ERROR'
    | 'UPLOAD_COMPLETE'
    | 'UPLOAD_ERROR'

export type FileItemT = {
    id: string
    name: string
    file?: File // only items that are not uploaded to s3 have this
    key?: string // only items uploaded to s3 have this
    s3URL?: string // only items uploaded to s3 have this
    status: FileStatus
}

const DocumentError = ({
    hasDuplicateNameError,
    hasScanningError,
    hasUploadError,
    hasUnexpectedError,
}: {
    hasDuplicateNameError: boolean
    hasScanningError: boolean
    hasUploadError: boolean
    hasUnexpectedError: boolean
}): React.ReactElement | null => {
    if (hasDuplicateNameError)
        return (
            <>
                <span className={styles.fileItemBoldMessage}>
                    Duplicate file
                </span>
                <span className={styles.fileItemBoldMessage}>
                    Please remove
                </span>
            </>
        )
    else if (hasScanningError && !hasUnexpectedError)
        return (
            <>
                <span className={styles.fileItemBoldMessage}>
                    Failed security scan, please remove
                </span>
            </>
        )
    else if (hasUploadError && !hasUnexpectedError)
        return (
            <>
                <span className={styles.fileItemBoldMessage}>
                    Upload failed
                </span>
                <span className={styles.fileItemBoldMessage}>
                    Please remove or retry
                </span>
            </>
        )
    else if (hasUnexpectedError) {
        return (
            <>
                <span className={styles.fileItemBoldMessage}>
                    Upload failed
                </span>
                <span className={styles.fileItemBoldMessage}>
                    Unexpected error. Please remove.
                </span>
            </>
        )
    } else {
        return null
    }
}

type FileItemProps = {
    item: FileItemT
    deleteItem: (item: FileItemT) => void
    retryItem: (item: FileItemT) => void
}
export const FileItem = ({
    item,
    deleteItem,
    retryItem,
}: FileItemProps): React.ReactElement => {
    const { name, status, file } = item
    const hasDuplicateNameError = status === 'DUPLICATE_NAME_ERROR'
    const hasScanningError = status === 'SCANNING_ERROR'
    const hasUploadError = status === 'UPLOAD_ERROR'
    const hasUnexpectedError = status === 'UPLOAD_ERROR' && file === undefined
    const isLoading = status === 'PENDING'
    const isScanning = status === 'SCANNING'

    const isPDF = name.indexOf('.pdf') > 0
    const isWord = name.indexOf('.doc') > 0 || name.indexOf('.pages') > 0
    const isVideo = name.indexOf('.mov') > 0 || name.indexOf('.mp4') > 0
    const isExcel = name.indexOf('.xls') > 0 || name.indexOf('.numbers') > 0
    const isGeneric = !isPDF && !isWord && !isVideo && !isExcel

    const imageClasses = classnames('usa-file-input__preview-image', {
        'is-loading': isLoading || isScanning,
        'usa-file-input__preview-image--pdf': isPDF,
        'usa-file-input__preview-image--word': isWord,
        'usa-file-input__preview-image--video': isVideo,
        'usa-file-input__preview-image--excel': isExcel,
        'usa-file-input__preview-image--generic': isGeneric,
    })

    const handleDelete = (_e: React.MouseEvent) => {
        deleteItem(item)
    }

    const handleRetry = (_e: React.MouseEvent) => {
        retryItem(item)
    }

    return (
        <>
            <div className={styles.fileItemText}>
                <img
                    id={item.id}
                    data-testid="file-input-preview-image"
                    src={SPACER_GIF}
                    alt=""
                    className={imageClasses}
                />
                <span
                    style={{
                        display: 'flex',
                        flexDirection: 'column',
                        fontSize: 'inherit',
                    }}
                >
                    <DocumentError
                        hasDuplicateNameError={hasDuplicateNameError}
                        hasScanningError={hasScanningError}
                        hasUploadError={hasUploadError}
                        hasUnexpectedError={hasUnexpectedError}
                    />
                    <>
                        {(isLoading || isScanning) && (
                            <span className={styles.fileItemBoldMessage}>
                                {isLoading
                                    ? 'Step 1 of 2: Uploading'
                                    : 'Step 2 of 2: Scanning'}
                            </span>
                        )}
                        <span>{name}</span>
                    </>
                </span>
            </div>
            <div className={styles.fileItemButtons}>
                <Button
                    type="button"
                    size="small"
                    unstyled
                    onClick={handleDelete}
                >
                    Remove
                </Button>
                {(hasUploadError || hasScanningError) && !hasUnexpectedError && (
                    <Button
                        type="button"
                        size="small"
                        unstyled
                        onClick={handleRetry}
                    >
                        Retry
                    </Button>
                )}
            </div>
        </>
    )
}
