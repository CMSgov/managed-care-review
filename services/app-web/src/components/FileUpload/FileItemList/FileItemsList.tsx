import React from 'react'
import classnames from 'classnames'
import { Table } from '@trussworks/react-uswds'
import { FileItem, FileItemT, FileStatus } from '../FileItem/FileItem'
import styles from '../FileUpload.module.scss'

export const FileItemsList = ({
    fileItems,
    deleteItem,
    retryItem,
    renderMode,
}: {
    fileItems: FileItemT[]
    deleteItem: (id: FileItemT) => void
    retryItem: (item: FileItemT) => void
    renderMode: 'table' | 'list'
}): React.ReactElement => {
    const liClasses = (status: FileStatus): string => {
        const hasError =
            status === 'UPLOAD_ERROR' ||
            status === 'SCANNING_ERROR' ||
            status === 'DUPLICATE_NAME_ERROR'
        return classnames(styles.fileItem, {
            'bg-secondary-lighter border-secondary margin-top-1px': hasError,
            'usa-file-input__preview': !hasError,
        })
    }

    if (renderMode === 'table') {
        return (
            <Table fullWidth>
                <thead>
                    <tr>
                        <th>Document name</th>
                        <th>Date uploaded</th>
                        <th></th>
                    </tr>
                </thead>
                <tbody>
                    {fileItems.map((item) => (
                        <FileItem
                            key={item.id}
                            deleteItem={deleteItem}
                            retryItem={retryItem}
                            item={item}
                            renderMode="table"
                        />
                    ))}
                </tbody>
            </Table>
        )
    } else {
        return (
            <ul
                data-testid="file-input-preview-list"
                className={styles.fileItemsList}
            >
                {fileItems.map((item) => (
                    <li
                        key={item.id}
                        id={item.id}
                        className={liClasses(item.status)}
                    >
                        <FileItem
                            deleteItem={deleteItem}
                            retryItem={retryItem}
                            item={item}
                            renderMode="list"
                        />
                    </li>
                ))}
            </ul>
        )
    }
}
