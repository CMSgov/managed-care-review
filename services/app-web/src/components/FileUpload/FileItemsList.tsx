import React from 'react'
import classnames from 'classnames'
import styles from './FileUpload.module.scss'
import { FileItem, FileItemT, FileStatus } from './FileItem'

export const FileItemsList = ({
    fileItems,
    deleteItem,
    retryItem,
}: {
    fileItems: FileItemT[]
    deleteItem: (id: FileItemT) => void
    retryItem: (item: FileItemT) => void
}): React.ReactElement => {
    const liClasses = (status: FileStatus): string => {
        const hasError =
            status === 'UPLOAD_ERROR' || status === 'DUPLICATE_NAME_ERROR'
        return classnames(styles.fileItem, {
            'bg-secondary-lighter border-secondary margin-top-1px': hasError,
            'usa-file-input__preview': !hasError,
        })
    }

    return (
        <ul
            data-testid="file-input-preview-list"
            className={styles.fileItemList}
        >
            {fileItems.map((item) => (
                <li
                    key={item.id}
                    id={item.id}
                    className={liClasses(item.status)}
                >
                    <FileItem
                        deleteItem={deleteItem}
                        retryItem={retryItem}
                        item={item}
                    />
                </li>
            ))}
        </ul>
    )
}
