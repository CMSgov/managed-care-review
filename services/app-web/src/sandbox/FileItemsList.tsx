import React from 'react'
import { FileItem, FileItemT } from './FileItem'

export const FileItemsList = ({
    fileItems,
    deleteItem,
}: {
    fileItems: FileItemT[]
    deleteItem: (id: string) => void
}): React.ReactElement => {
    return (
        <ul
            style={{
                listStyleType: 'none',
                display: 'inline-block',
                padding: 0,
                margin: '0 0 -1px ',
                width: '480px',
            }}
        >
            {fileItems.map((item) => (
                <li
                    key={item.id}
                    id={item.id}
                    className={
                        item.status === 'UPLOAD_ERROR'
                            ? 'bg-secondary-lighter border-secondary'
                            : 'usa-file-input__preview'
                    }
                    style={{
                        display: 'flex',
                        justifyContent: 'space-between',
                        pointerEvents: 'all',
                    }}
                >
                    <FileItem deleteItem={deleteItem} item={item} />
                </li>
            ))}
        </ul>
    )
}
