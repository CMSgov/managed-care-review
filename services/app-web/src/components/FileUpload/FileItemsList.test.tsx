import React from 'react'
import { render, screen } from '@testing-library/react'

import { FileItemT } from './FileItem'
import { FileItemsList } from './FileItemsList'

describe('FileItem component', () => {
    const testProps = {
        fileItems: [
            {
                id: 'testFile1',
                name: 'testFile1.pdf',
                url: undefined,
                key: undefined,
                s3URL: undefined,
                status: 'PENDING',
            },
            {
                id: 'testFile2',
                name: 'testFile2.pdf',
                url: undefined,
                key: undefined,
                s3URL: undefined,
                status: 'UPLOAD_ERROR',
            },
        ] as FileItemT[],
        deleteItem: (fileItem: FileItemT) => {
            /* delete item*/
        },
        retryItem: (fileItem: FileItemT) => {
            /* retry item*/
        },
    }

    it('renders without errors', () => {
        render(<FileItemsList {...testProps} />)

        expect(screen.getAllByRole('listitem').length).toEqual(
            testProps.fileItems.length
        )
        expect(screen.getByText('testFile1.pdf')).toBeInTheDocument()
        expect(screen.getByText('testFile2.pdf')).toBeInTheDocument()
    })

    it('displays error styles for items that have errors', () => {
        render(<FileItemsList {...testProps} />)
        const listItems = screen.getAllByRole('listitem')
        expect(listItems[0]).toHaveClass('usa-file-input__preview')
        expect(listItems[1]).not.toHaveClass('usa-file-input__preview')
        expect(listItems[1]).toHaveClass(
            'bg-secondary-lighter border-secondary '
        )
    })
})
