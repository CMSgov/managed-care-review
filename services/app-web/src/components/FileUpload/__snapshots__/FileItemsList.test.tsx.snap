// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileItem component renders correctly 1`] = `
<ul
  className="fileItemsList"
  data-testid="file-input-preview-list"
>
  <li
    className="fileItem usa-file-input__preview"
    id="testFile1"
  >
    <div
      className="fileItemText"
    >
      <img
        alt=""
        className="usa-file-input__preview-image is-loading usa-file-input__preview-image--pdf"
        data-testid="file-input-preview-image"
        id="testFile1"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      />
      <span
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "fontSize": "inherit",
          }
        }
      >
        <span>
          testFile1.pdf
        </span>
      </span>
    </div>
    <div
      className="fileItemButtons"
    >
      <button
        className="usa-button usa-button--small usa-button--unstyled"
        data-testid="button"
        onClick={[Function]}
        type="button"
      >
        Remove
      </button>
    </div>
  </li>
  <li
    className="fileItem bg-secondary-lighter border-secondary margin-top-1px"
    id="testFile2"
  >
    <div
      className="fileItemText"
    >
      <img
        alt=""
        className="usa-file-input__preview-image usa-file-input__preview-image--pdf"
        data-testid="file-input-preview-image"
        id="testFile2"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      />
      <span
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "fontSize": "inherit",
          }
        }
      >
        <span
          className="fileItemErrorMessage"
        >
          Upload failed
        </span>
        <span
          className="fileItemErrorMessage"
        >
          Please remove or retry
        </span>
        <span>
          testFile2.pdf
        </span>
      </span>
    </div>
    <div
      className="fileItemButtons"
    >
      <button
        className="usa-button usa-button--small usa-button--unstyled"
        data-testid="button"
        onClick={[Function]}
        type="button"
      >
        Remove
      </button>
      <button
        className="usa-button usa-button--small usa-button--unstyled"
        data-testid="button"
        onClick={[Function]}
        type="button"
      >
        Retry
      </button>
    </div>
  </li>
</ul>
`;
